### PR TITLE
Gemspecs cannot have dynamic dependencies in them

### PR DESCRIPTION
**Pull request by bf4**

### CODE REVIEW

**Comment by bf4** [comment]
Transient failure 

- https://travis-ci.org/sensu/sensu/jobs/79522897

```plain
  1) Sensu::Server::Process can consume results
     Failure/Error: expect(event[:occurrences]).to eq(2)
       
       expected: 2
            got: 1
       
       (compared using ==)
     # ./spec/server/process_spec.rb:174:in `block (11 levels) in <top (required)>'
     # ./spec/helpers.rb:57:in `async_wrapper'
     # ./spec/server/process_spec.rb:152:in `block (2 levels) in <top (required)>'
```

**Comment by bf4** [comment]
- https://travis-ci.org/sensu/sensu/jobs/79522897

```plain

  1) Sensu::Server::Process can consume results
     Failure/Error: expect(event[:occurrences]).to eq(2)
       
       expected: 2
            got: 1
       
       (compared using ==)
     # ./spec/server/process_spec.rb:174:in `block (11 levels) in <top (required)>'
     # ./spec/helpers.rb:57:in `async_wrapper'
     # ./spec/server/process_spec.rb:152:in `block (2 levels) in <top (required)>'
```

**Comment by portertech** [comment]
Why not?

**Comment by bf4** [comment]
Because the gemspec is only evaluated when the gem is built.  If you're building a fat gems for multiple platforms, then you need to specify the current platform and build it for each one, but they'll different names.

**Comment by bf4** [comment]
Its meaningless to say `s.add_dependency "json" if RUBY_VERSION < "1.9"` when you're building the gem on 1.9 or higher.   If it's built on < 1.9, then it's meaningless to have the condition.

Similarly, `s.add_dependency "thin", "1.6.3" unless RUBY_PLATFORM =~ /java/` is only evaluated when the gem is built, so if you're building it with cruby, this dep will included in the gem, on JRuby it won't be, but they'll overwrite each other and you'll only be able to push once.

**Comment by bf4** [comment]
You're an author and owner of the gem, so I'm presuming you may have good reasons for doing it this way, but I can't deduce them from the code.

**Comment by bf4** [comment]
While I'm commenting here, I made some other small plumbing prs and issue

- [Issues](https://github.com/sensu/sensu/issues/created_by/bf4)
  - [Use appveyor to test on Windows](https://github.com/sensu/sensu/issues/1049)
- [PRs](https://github.com/sensu/sensu/pulls/bf4)
  - [Loosen EventMachine dependency, except on Windows](https://github.com/sensu/sensu/pull/1045)
  - [Update executable convention, avoids conflict with binstubs](https://github.com/sensu/sensu/pull/1047)

I'll rebase any with conflicts that you're interested in.

If you have a gitter/slack/irc we can chat there, as well.

**Comment by portertech** [comment]
@bf4 The reason the gemspec contains conditionals is for building gems that target different platforms, CRuby vs. JRuby, as certain gems cannot build on one platform or the other.

File: sensu.gemspec [hunk 1]
```diff
@@ -27,7 +27,7 @@ Gem::Specification.new do |s|
   s.add_dependency "em-redis-unified", "1.0.0"
   s.add_dependency "sinatra", "1.4.6"
   s.add_dependency "async_sinatra", "1.2.0"
-  s.add_dependency "thin", "1.6.3" unless RUBY_PLATFORM =~ /java/
```
Comment (portertech) [inline sensu.gemspec]: Thin does not build on JRuby (c bindings).

File: sensu.gemspec [hunk 2]
```diff
@@ -4,7 +4,7 @@ require File.join(File.dirname(__FILE__), "lib", "sensu", "constants")
 Gem::Specification.new do |s|
   s.name        = "sensu"
   s.version     = Sensu::VERSION
-  s.platform    = RUBY_PLATFORM =~ /java/ ? Gem::Platform::JAVA : Gem::Platform::RUBY
+  s.platform    = Gem::Platform::CURRENT
```
Comment (portertech) [inline sensu.gemspec]: This should work when building on both platforms :+1:

File: sensu.gemspec [hunk 3]
```diff
@@ -13,7 +13,7 @@ Gem::Specification.new do |s|
   s.license     = "MIT"
   s.has_rdoc    = false
 
-  s.add_dependency "json" if RUBY_VERSION < "1.9"
```
Comment (portertech) [inline sensu.gemspec]: Removing this conditional is probably OK, as JSON is included in STDLIB > 1.8

File: sensu.gemspec (hunk 1 above)
Comment (bf4) [inline sensu.gemspec]: Ah, I apologize that I missed you're building a java-specific gem

- https://rubygems.org/gems/sensu/versions/0.20.5-java
- https://rubygems.org/gems/sensu/versions/0.20.5

I expected to see something in the source code (`ext/sensu/extconf.rb`) and didn't think to look at RubyGems.

Not sure what your build process is. Are you using https://github.com/rake-compiler/rake-compiler-dock or https://github.com/tjschuck/rake-compiler-dev-box ( https://github.com/rake-compiler/rake-compiler )? I love that workflow.

Then you could just modify the thin dep in your buildstep, e.g. https://github.com/codahale/bcrypt-ruby/blob/1f9184a8df2b90fa02d01a32a364eefc9072e1b9/Rakefile#L49-L59

**Comment by portertech** [comment]
Going to close this, as the current gemspec works with the sensu-build pipeline. We can revisit/reopen this if the build pipeline is impacted. Thank you for submitting the PR! :+1:

**Comment by bf4** [comment]
Let me know if I can help in anyway. You might also want to consider generating the gemspec from your rakefile perhaps via hoe, to handle platform different gemspecs in a more conventional way

B mobile phone

**Pull request closed without merging**